From e44cf26ed3078947cfda912dc4f5cf9ff1147933 Mon Sep 17 00:00:00 2001
From: Prarit Bhargava <prarit@redhat.com>
Date: Thu, 30 Oct 2014 18:57:16 -0400
Subject: [x86] smpboot: Fix up typo in topology detection

Message-id: <1414695436-7361-1-git-send-email-prarit@redhat.com>
Patchwork-id: 98590
O-Subject: [RHEL7.1 PATCH BZ 1156655] sched/x86: Fix up typo in topology detection
Bugzilla: 1156655
RH-Acked-by: Larry Woodman <lwoodman@redhat.com>
RH-Acked-by: Igor Mammedov <imammedo@redhat.com>
RH-Acked-by: Don Zickus <dzickus@redhat.com>

Bugzilla: https://bugzilla.redhat.com/show_bug.cgi?id=1156655
Build info:
Date: Thu Oct 30 13:22:37 EDT 2014
Build OS: Red Hat Enterprise Linux Server release 7.0 (Maipo)
System name: intel-canoepass-05.lab.bos.redhat.com with -j24
Built on: kernel-3.10.0-196.el7
Arch built: ppc64 s390 x86_64

This patch explicitly depends on the patches for BZ 1158269 posted earlier
today.

commit 728e5653e6fdb2a0892e94a600aef8c9a036c7eb
Author: Dave Hansen <dave.hansen@linux.intel.com>
Date:   Tue Sep 30 14:45:46 2014 -0700

    sched/x86: Fix up typo in topology detection

    Commit:

      cebf15eb09a2 ("x86, sched: Add new topology for multi-NUMA-node CPUs")

    some code to try to detect the situation where we have a NUMA node
    inside of the "DIE" sched domain.

    It detected this by looking for cpus which match_die() but do not match
    NUMA nodes via topology_same_node().

    I wrote it up as:

     if (match_die(c, o) == !topology_same_node(c, o))

    which actually seemed to work some of the time, albiet
    accidentally.

    It should have been doing an &&, not an ==.

    This code essentially chopped off the "DIE" domain on one of
    Andrew Morton's systems.  He reported that this patch fixed his
    issue.

    Signed-off-by: Dave Hansen <dave.hansen@linux.intel.com>
    Reported-by: Andrew Morton <akpm@linux-foundation.org>
    Signed-off-by: Peter Zijlstra (Intel) <peterz@infradead.org>
    Cc: Dave Hansen <dave@sr71.net>
    Cc: Andrew Morton <akpm@linux-foundation.org>
    Cc: David Rientjes <rientjes@google.com>
    Cc: Igor Mammedov <imammedo@redhat.com>
    Cc: Jan Kiszka <jan.kiszka@siemens.com>
    Cc: Lan Tianyu <tianyu.lan@intel.com>
    Cc: Linus Torvalds <torvalds@linux-foundation.org>
    Cc: Prarit Bhargava <prarit@redhat.com>
    Cc: Toshi Kani <toshi.kani@hp.com>
    Link: http://lkml.kernel.org/r/20140930214546.FD481CFF@viggo.jf.intel.com
    Signed-off-by: Ingo Molnar <mingo@kernel.org>

Cc: Igor Mammedov <imammedo@redhat.com>
Cc: Tony Camuso <tcamuso@redhat.com>
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/arch/x86/kernel/smpboot.c b/arch/x86/kernel/smpboot.c
index 5ebce78..176846b 100644
--- a/arch/x86/kernel/smpboot.c
+++ b/arch/x86/kernel/smpboot.c
@@ -412,7 +412,7 @@ void set_cpu_sibling_map(int cpu)
    } else if (i != cpu && !c->booted_cores)
     c->booted_cores = cpu_data(i).booted_cores;
   }
-  if (match_die(c, o) == !topology_same_node(c, o))
+  if (match_die(c, o) && !topology_same_node(c, o))
    primarily_use_numa_for_topology();
  }
 }
-- 
1.7.1